File: Samples/visio_managed_code_interop_vs2010/Visio2010_VB_Samples/Page_GetFormulas.vb
Imports IVisio = Microsoft.Office.Interop.Visio


Partial Public Class VS2010_VB_Samples
    Shared Sub Page_GetFormulas(ByVal doc As Microsoft.Office.Interop.Visio.Document)
        Dim page = VisioInterop.Util.CreateStandardPage(doc, "PGF")
        Dim shape = VisioInterop.Util.CreateStandardShape(page)

        ' CREATE REQUEST
        Dim request = {New With { _
                Key .ID = CShort(shape.ID16), _
                Key .Section = CShort(IVisio.VisSectionIndices.visSectionObject), _
                Key .Row = CShort(IVisio.VisRowIndices.visRowXFormOut), _
                Key .Cell = CShort(IVisio.VisCellIndices.visXFormWidth) _
                }, New With { _
                Key .ID = CShort(shape.ID16), _
                Key .Section = CShort(IVisio.VisSectionIndices.visSectionObject), _
                Key .Row = CShort(IVisio.VisRowIndices.visRowXFormOut), _
                Key .Cell = CShort(IVisio.VisCellIndices.visXFormHeight) _
                }}

        ' MAP THE REQUEST TO THE STRUCTURES VISIO EXPECTS
        Dim SID_SRCStream = New Short(request.Length * 4 - 1) {}
        For i As Integer = 0 To request.Length - 1
            SID_SRCStream((i * 4) + 0) = request(i).ID
            SID_SRCStream((i * 4) + 1) = request(i).Section
            SID_SRCStream((i * 4) + 2) = request(i).Row
            SID_SRCStream((i * 4) + 3) = request(i).Cell
        Next

        ' EXECUTE THE REQUEST
        Dim formulas_sa As System.Array
        page.GetFormulasU(SID_SRCStream, formulas_sa)

        ' MAP OUTPUT BACK TO SOMETHING USEFUL 
        Dim formulas = New String(formulas_sa.Length - 1) {}
        formulas_sa.CopyTo(formulas, 0)

        ' DISPLAY THE INFORMATION
        shape.Text = String.Format("Formulas={0},{1}", formulas(0), formulas(1))

    End Sub
End Class

File: Samples/visio_managed_code_interop_vs2010/VS1010_CSharp_Samples/page_getformulas.cs
using VisioInterop;
using IVisio = Microsoft.Office.Interop.Visio;
using System.Collections.Generic;

public static partial class VS2010_CSharp_Samples
{
    public static void Page_GetFormulas(IVisio.Document doc)
    {
        var page = VisioInterop.Util.CreateStandardPage(doc, "PGF");
        var shape = VisioInterop.Util.CreateStandardShape(page);
        var request = Util.Create_PageGetFormulas_Request(shape);

        // MAP THE REQUEST TO THE STRUCTURES VISIO EXPECTS
        var SID_SRCStream = new short[request.Length*4];
        for (int i = 0; i < request.Length; i++)
        {
            SID_SRCStream.Set4(i, request[i].ShapeID, request[i].CellSRC.SectionIndex, request[i].CellSRC.RowIndex, request[i].CellSRC.CellIndex);
        }


        // EXECUTE THE REQUEST
        System.Array formulas_sa;
        page.GetFormulasU(SID_SRCStream, out formulas_sa);

        // MAP OUTPUT BACK TO SOMETHING USEFUL 
        var formulas = new string[formulas_sa.Length];
        formulas_sa.CopyTo(formulas, 0);

        // DISPLAY THE INFORMATION
        shape.Text = string.Format("Formulas={0},{1}", formulas[0], formulas[1]);
    }
}